src/TesApi.Web/CachingAzureProxy.cs

```diff
@@ -0,0 +1,136 @@
+        public int GetBatchActiveJobCount() => azureProxy.GetBatchActiveJobCount();
+        public IEnumerable<AzureProxy.AzureBatchNodeCount> GetBatchActiveNodeCountByVmSize() => azureProxy.GetBatchActiveNodeCountByVmSize();
+        public int GetBatchActivePoolCount() => azureProxy.GetBatchActivePoolCount();
+        public Task<AzureProxy.AzureBatchJobAndTaskState> GetBatchJobAndTaskStateAsync(string tesTaskId) => azureProxy.GetBatchJobAndTaskStateAsync(tesTaskId);
+
+        ///<inheritdoc/>
+        public async Task<ContainerRegistryInfo> GetContainerRegistryInfoAsync(string imageName)
+        {
+            if (!cache.TryGetValue(imageName, out ContainerRegistryInfo containerRegistry))
+            {
+                // retry
+                containerRegistry = await azureProxy.GetContainerRegistryInfoAsync(imageName);
+
+                if (containerRegistry != null)
+                {
+                    cache.Set(imageName, containerRegistry, TimeSpan.FromHours(1));
+                }
+            }
+
+            return containerRegistry;
+        }
+
+        public Task<string> GetNextBatchJobIdAsync(string tesTaskId) => azureProxy.GetNextBatchJobIdAsync(tesTaskId);
+        public Task<IEnumerable<string>> GetPoolIdsReferencedByJobsAsync(CancellationToken cancellationToken) => azureProxy.GetPoolIdsReferencedByJobsAsync(cancellationToken);
+
+        ///<inheritdoc/>
+        public async Task<string> GetStorageAccountKeyAsync(StorageAccountInfo storageAccountInfo)
+        {
+            if (!cache.TryGetValue(storageAccountInfo.Id, out string accountKey))
+            {
+                // retry
+                accountKey = await azureProxy.GetStorageAccountKeyAsync(storageAccountInfo);
+                cache.Set(storageAccountInfo.Id, accountKey, TimeSpan.FromHours(1));
+            }
+
+            return accountKey;
+        }
+
+        ///<inheritdoc/>
+        public async Task<StorageAccountInfo> GetStorageAccountInfoAsync(string storageAccountName)
+        {
+            if (!cache.TryGetValue(storageAccountName, out StorageAccountInfo storageAccountInfo))
+            {
+                // retry
+                storageAccountInfo = await azureProxy.GetStorageAccountInfoAsync(storageAccountName);
+
+                if (storageAccountInfo != null)
+                {
+                    cache.Set(storageAccountName, storageAccountInfo, DateTimeOffset.MaxValue);
+                }
+            }
+
+            return storageAccountInfo;
+        }
+
+        ///<inheritdoc/>
+        public async Task<List<VirtualMachineInfo>> GetVmSizesAndPricesAsync()
+        {
+            const string key = "vmSizesAndPrices";
+
+            if (!cache.TryGetValue(key, out List<VirtualMachineInfo> cachedVmSizesAndPrices))
+            {
+                // retry
+                var vmSizesAndPrices = await azureProxy.GetVmSizesAndPricesAsync();
+                cache.Set(key, vmSizesAndPrices.ToList(), DateTimeOffset.MaxValue);
```

Comment:
This means that VM sizes and prices are effectively only updated if the VM is restarted, which means that the data could be stale by over a month.  Perhaps cache for (1) day?

Reply:
Since the API call takes a while, @tonybendis and I discussed adding a background task that updates the cache periodically instead without adding as much delay. Tracking now #78